### PR TITLE
gardener-apiserver: Improve the HPA -> VPAAndHPA transition

### PR DESCRIPTION
**Pull request by ialidzhikov**
**How to categorize this PR?**

/area auto-scaling
/kind enhancement

**What this PR does / why we need it**:
See https://github.com/gardener/gardener/blob/d744f46804ebe36ef944b5de2ece32950260d007/pkg/component/gardener/apiserver/deployment.go#L151-L161

**Which issue(s) this PR fixes**:
Follow-up after #9735 

**Special notes for your reviewer**:
The HPA -> VPAAndHPA transition for gardener-apiserver transition is not perfect as step 2. can also decrease the gardener-apiserver Deployment replicas to 2, set by the GRM's HA webhook because at that time there is no HPA object yet.  Then the new HPA gets created and for HA Garden cluster, it is created with minReplicas=3. There can be also a change in the Pod template in the Deployment - in the minDomains of the TSC which is computed as `min(maxReplicas, numberOfZones)`.

At least, this PR improves from Deployment resources aspect and makes sure that the new gardener-apiserver Pods won't be created with the initial/default Deployment resource requests.

**Release note**:

```other operator
gardener-operator: The HVPA -> VPAAndHHPA autoscaling mode transition is now improved for the gardener-apiserver to preserve the Deployment resources.
```

### CODE REVIEW

File: pkg/component/gardener/apiserver/deployment.go [hunk 1]
```diff
@@ -147,5 +147,21 @@ func (g *gardenerAPIServer) deployment(
...
+	// However, druing the HPA -> VPAAndHPA transition the flow is as follows:
+	// 1. The HVPA resource is deleted.
+	// 2. The Deployment is updated.
+	// 3. The new HPA and VPA resources are created.
+	//
+	// Step 2. is the one that reverts the Deployment resources to the initial ones as during that time there is no existing HVPA object.
+	//
+	// TODO(ialidzhikov): Drop this if statement when the VPAAndHPA is the only one supported autoscaling mode.
+	if g.values.Autoscaling.Mode == apiserver.AutoscalingModeVPAAndHPA {
+		metav1.SetMetaDataAnnotation(&deployment.ObjectMeta, resourcesv1alpha1.PreserveResources, "true")
```
Comment (voelzmo) [inline pkg/component/gardener/apiserver/deployment.go:163]: While we're doing this: do we also need to set `resourcesv1alpha1.PreserveReplicas` here?

**Review by voelzmo — COMMENTED** [review]
Looks good in general, just a question regarding modification of replicas.

File: pkg/component/gardener/apiserver/deployment.go (hunk 1 above)
Comment (ialidzhikov) [reply pkg/component/gardener/apiserver/deployment.go:163]: Hi @voelzmo ,

I played with the PreserveReplicas annotation on Friday and I tried to explain the outcome in the `Special notes for your reviewer` section.
Even if GRM preserves the replicas when updating the Deployment, the replicas are always set by the highavailabilityconfig webhook of the GRM. See https://github.com/gardener/gardener/blob/db182529c3c1badd987f7e5b20bdcc71f33f8e0a/pkg/resourcemanager/webhook/highavailabilityconfig/handler.go#L313-L318
During the migration HVPA -> VPAAndHPA, HVPA will be deleted first, then the Deployment gets updated. Hence, `isHorizontallyScaled` will be false and the HA config webhook of GRM will mutate the replicas.

tl;dr: According to my testing setting `resourcesv1alpha1.PreserveReplicas` does not have effect in this particular case.

Comment (voelzmo) [reply pkg/component/gardener/apiserver/deployment.go:163]: Ah, thanks for explaining it in a bit more detail. I had read this section, but didn't really understand that the annotation doesn't apply in this case. I get now that this is due to the flow of deleting the HPA object, updating Deployment (and thereby HA config webhook setting replicas to 2) and only then creating a new HPA object.
If we really want, we could avoid this by setting `high-availability-config.resources.gardener.cloud/replicas: 3`, not sure if this is really necessary as this will anyways happen when the HPA gets created with `minReplicas: 3`?

Comment (ialidzhikov) [reply pkg/component/gardener/apiserver/deployment.go:163]: Hm, IIUC the proposal is to make use of `high-availability-config.resources.gardener.cloud/replicas: 3`. According to https://github.com/gardener/gardener/blob/450529c84b99ddc36a3ec43abb1e530839fc18a5/docs/concepts/resource-manager.md#L964-L967, the HA webhook of GRM does NOT mutate the replicas
> when the resource is scaled horizontally by `HorizontalPodAutoscaler` or `Hvpa`, and the current replica count is higher than what was computed above.

Hence, `high-availability-config.resources.gardener.cloud/replicas: 3` will be used only during the migration from HVPA to HPAAndVPA autoscaling mode. Then, after the HPA is created for the gardener-apiserver, HPA will control the replicas count, the HA webhook won't overwrite the replicas count and the `high-availability-config.resources.gardener.cloud/replicas` annotation on the Deployment won't be respected.

It looks hacky but I can test and if it works, it improves a little bit the transition.

Comment (ialidzhikov) [reply pkg/component/gardener/apiserver/deployment.go:163]: I played with the above proposal in local setup. For some reason that I didn't fully investigate, even with `high-availability-config.resources.gardener.cloud/replicas: 3` I see minDomains flapping between 2 <-> 3 during the migration from HVPA mode with currentReplicas=4,minReplicas=4,maxReplicas=4 (just for testing I adapted the HVPA to run with 4 replicas for a more real world testing) and VPAAndMode minReplicas=3,maxReplicas=6.

It is not super clear proposal but rather relying on implicit contracts. And at the end it does not avoid the new ReplicaSet creation (due this minDomains explained in the above sentence). So, I would rather leave it as it is and accept the 4 -> 2 scale down, then HPA scale up to 3 replicas and then the new ReplicaSet due to the minDomains change. At the end, we will have at least 2 running replicas during the migration so it is not super bad - just fluctuation/noise during the migration which should stabilize with the new ReplicaSet.

File: pkg/component/gardener/apiserver/deployment.go [hunk 2]
```diff
@@ -147,5 +147,21 @@ func (g *gardenerAPIServer) deployment(
 
 	utilruntime.Must(gardenerutils.InjectGenericKubeconfig(deployment, secretGenericTokenKubeconfig.Name, secretVirtualGardenAccess.Secret.Name))
 	utilruntime.Must(references.InjectAnnotations(deployment))
+
+	// Preserve the Deployment resources during the HPA -> VPAAndHPA transition.
+	//
+	// The gardener-apiserver is deployed via ManagedResource. In the general case, GRM preserves the Deployment resources when the resource is being scaled by HVPA.
+	// However, druing the HPA -> VPAAndHPA transition the flow is as follows:
```
Comment (oliver-goetz) [inline pkg/component/gardener/apiserver/deployment.go]:
```suggestion
	// However, during the HPA -> VPAAndHPA transition the flow is as follows:
```
nit

**Review by oliver-goetz — COMMENTED** [review]
There is one typo, otherwise it looks good 🙂

**Comment by oliver-goetz** [comment]
/approve

@voelzmo could you please check and lgtm? 🙂

**Review by timuthy — COMMENTED** [review]
/lgtm
